src/main/java/jenkins/plugins/git/GitRepoSizeEstimator.java

```diff
@@ -0,0 +1,184 @@
+        boolean useCache;
+        String cacheEntry = source.getCacheEntry();
+        File cacheDir = AbstractGitSCMSource.getCacheDir(cacheEntry);
+        if (cacheDir != null) {
+            Git git = Git.with(TaskListener.NULL, new EnvVars(EnvVars.masterEnvVars)).in(cacheDir).using("git");
+            GitClient client = git.getClient();
+            if (!client.hasGitRepo()) {
+                useCache = false;
+            } else {
+                useCache = true;
+                sizeOfRepo = FileUtils.sizeOfDirectory(cacheDir);
+                sizeOfRepo = (sizeOfRepo/1000); // Conversion from Bytes to Kilo Bytes
+            }
+        } else {
+            useCache = false;
+        }
+        return useCache;
+    }
+
+    /**
+     * Check if the desired implementation of extension is present and ask for the size of repository if it does
+     * @param repoUrl: The remote name derived from {@link GitSCMSource} object
+     * @return boolean useAPI or not.
+     */
+    private boolean setSizeFromAPI(String repoUrl) {
+        List<RepositorySizeAPI> acceptedRepository = Objects.requireNonNull(RepositorySizeAPI.all())
+                .stream()
+                .filter(r -> r.acceptsRemote(repoUrl))
+                .collect(Collectors.toList());
+
+        if (acceptedRepository.size() == 1) {
+            sizeOfRepo = acceptedRepository.get(0).getSizeOfRepository(repoUrl);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Recommend a git implementation on the basis of the given size of a repository
+     * @param sizeOfRepo: Size of a repository (in KiBs)
+     * @return a git implementation, "git" or "jgit"
+     */
+    private String determineSwitchOnSize(Long sizeOfRepo) {
+        if (sizeOfRepo != 0L) {
+            if (sizeOfRepo >= SIZE_TO_SWITCH) {
+                return "git";
+            } else {
+                return "jgit";
+            }
+        }
+        return "NONE";
+    }
+
+    /**
+     * For a given recommended git implementation, validate if the installation exists and provide no suggestion if
+     * implementation doesn't exist.
+     * @param gitImplementation: The recommended git implementation, "git" or "jgit" on the basis of the heuristics.
+     */
+    private void determineGitTool(String gitImplementation) {
+        if (gitImplementation.equals("DEFAULT")) {
+            gitTool = "NONE";
+            return; // Recommend nothing (GitToolRecommendation = NONE)
+        }
+        final Jenkins jenkins = Jenkins.get();
+        GitTool tool = GitUtils.resolveGitTool(gitImplementation, jenkins, null, TaskListener.NULL);
+        if (tool != null) {
+            gitTool = tool.getGitExe();
+        }
+    }
+
+    /**
+     * Recommend git tool to be used by the git client
+     * @return git implementation recommendation in the form of a string
+     */
+    public String getGitTool() {
+        return gitTool;
+    }
```

Comment:
Recovering the comment missing after the last commit:
I was wondering if we just return the GitTool. What do you think @MarkEWaite ?